### ISSUE TITLE
Setting "evaluate_on_test", the last model is evaluated, not the best one

### ISSUE THREAD

**Issue by janLuke** [allenai/allennlp#1223]
While everyone would expect the best weights are used. The best weights are the output of the training process, not the last ones. I think the current behavior is potentially harmful because it misguides who reads the results taking for granted (almost surely) those are the results for the best model.

**Comment by matt-gardner** [comment]
Yep, definitely a bug.  PRs welcome.

**Comment by matt-gardner** [comment]
My vote would be for the best model, I think.

**Comment by joelgrus** [comment]
best model would fix #1192 as well

**Comment by matt-gardner** [comment]
Yes, all of that sounds like an improvement to me.  We've been more focused on defining model-building APIs than on the particulars of the `Trainer`, which is why it's not as fully-featured as you might like.  Again, PRs are greatly appreciated - we're a small team.
